src/compiler/moduleSpecifiers.ts

```diff
@@ -1026,6 +1028,46 @@ function tryGetModuleNameFromPackageJsonImports(moduleFileName: string, sourceDi
     })?.moduleFileToTry;
 }
 
+function tryGetModuleNameFromPackageJsonDependencies(moduleFileName: Path, { sourceDirectory, getCanonicalFileName }: Info, options: CompilerOptions, host: ModuleSpecifierResolutionHost, importMode: ResolutionMode) {
+    if (!host.readFile) {
+        return undefined;
+    }
+    const cache = host.getModuleResolutionCache?.();
+    if (!cache) {
+        return undefined;
+    }
+
+    const ancestorDirectoryWithPackageJson = getNearestAncestorDirectoryWithPackageJson(host, sourceDirectory);
+    if (!ancestorDirectoryWithPackageJson) {
+        return undefined;
+    }
+    const packageJsonPath = combinePaths(ancestorDirectoryWithPackageJson, "package.json");
+    const cachedPackageJson = host.getPackageJsonInfoCache?.()?.getPackageJsonInfo(packageJsonPath);
+    if (isMissingPackageJsonInfo(cachedPackageJson) || !host.fileExists(packageJsonPath)) {
+        return undefined;
+    }
+    const packageJsonContent = cachedPackageJson?.contents.packageJsonContent || tryParseJson(host.readFile(packageJsonPath)!);
+    const dependencies = packageJsonContent?.dependencies;
+    if (!dependencies) {
+        return undefined;
+    }
+
+    const candidates = getOwnKeys(dependencies);
+    for (const name of candidates) {
+        const resolved = resolveModuleName(name, combinePaths(sourceDirectory, "dummy.ts"), options, host as (typeof host & { readFile: NonNullable<(typeof host)["readFile"]> }), cache, /*redirectedReference*/ undefined, importMode);
+        if (resolved && resolved.resolvedModule && resolved.resolvedModule) {
+            const resolvedPath = toPath(resolved.resolvedModule.resolvedFileName, sourceDirectory, getCanonicalFileName);
+            if (resolvedPath === moduleFileName) {
+                return name;
+            }
+            if (host.realpath && host.realpath(resolvedPath) === host.realpath(moduleFileName)) {

```

Comment:
I didn't *want* to have to `realpath` here, but it seems to be the only way we can check if the two paths actually refer to the same thing through symlinks.